<h1 align=center><font size = 5>Final Project</font></h1>

## Idea:
-I have an Italian Investor who wants to open an Italian restaurant in New York and he is looking for the best neighborhood for his restaurant<br/>
-The idea is to find a neighborhood who doesn't have an  Italian Restaurant and is also quite central.<br/>
-His ideal neighborhood would be one which is central , doesn't have any Italian restaurant and accessible to everyone.<br/>

## Data:
-I have access to New York data which has the different neighborhoods and their longitude and latitude.<br/>
-First I will use this data with the FourSquare API to try to find the neighborhoods with Zero Italian Restaurants<br/>
-Then I will create a folium map of it and show it to the investor.<br/>


In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [7]:
#Get the Data
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')
    
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Data downloaded!


In [8]:
#Cleaning the data
neighborhoods_data = newyork_data['features']

# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

In [9]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [197]:
neighborhoods.head(2)

(306, 4)

## Second Step:
-After I got the dataset and cleaned it above, Now i will go through all the Borough and find the Neighborhood with the minimum number of Italian Restaurants.<br/>

In [198]:
CLIENT_ID = 'SL4ZRKTXASCIAW2GSIJ5X1OZJU0XFRS4KVW2VVBCIGXQILPU' # your Foursquare ID
CLIENT_SECRET = '5PRRPFCBUSQGQ4PAGGNLHH4UU3GWLAPWO2K3JEKRMS5IOR35' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
search_query = 'food'
LIMIT=100



# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['Venue Category']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']
    
#Get NearByFood places in all the neighborhoods in one Borough
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET,
            lat,
            lng,
            VERSION,
            search_query,
            radius,
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['venues']

         
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['name'], 
            v['location']['lat'], 
            v['location']['lng'],  
            v['categories']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue Name', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    # filter the category for each row
    nearby_venues['Venue Category'] = nearby_venues.apply(get_category_type, axis=1)
    
    return(nearby_venues)



In [ ]:
#Get the list of all food venues in all neighborhoods
All_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )



In [202]:
All_venues.head(6)

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3   Co-op City              40.874294              -73.829939   
4   Co-op City              40.874294              -73.829939   
5   Co-op City              40.874294              -73.829939   

                                 Venue Name  Venue Latitude  Venue Longitude  \
0              Baychester Avenue Food Truck       40.892293       -73.843230   
1  Taste the Island Bakery & Caribbean Food       40.899265       -73.846916   
2                  Foodtown White Plains Rd       40.890900       -73.851403   
3                 Food Universe Marketplace       40.876740       -73.828980   
4                 Food Universe Marketplace       40.870520       -73.828550   
5                             Compare Foods       40.870647       -73.828425   

       Venue Category  
0          Food Truck  
1                None  
2         Supermarket  
3       Grocery Store  
4         Supermarket  
5  Miscellaneous Shop

In [203]:
def get_neighborhoods_zero_italian_restaurants(venues):
    # one hot encoding
    venues_onehot = pd.get_dummies(venues[['Venue Category']], prefix="", prefix_sep="")

    # add Neighborhood Latitude back to dataframe
    venues_onehot['Neighborhood Latitude'] = venues['Neighborhood Latitude'] 
    
    # add Neighborhood Longitude back to dataframe
    venues_onehot['Neighborhood Longitude'] = venues['Neighborhood Longitude']     
    
    # add neighborhood column back to dataframe
    venues_onehot['Neighborhood'] = venues['Neighborhood'] 
    
    # move neighborhood column to the first column
    fixed_columns = [venues_onehot.columns[-1]] + list(venues_onehot.columns[:-1])
    venues_onehot = venues_onehot[fixed_columns]

    #Group By Neighborhood
    venues_onehot=venues_onehot.groupby(['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude']).sum()
    venues_onehot=venues_onehot.reset_index(drop = False)

    return (venues_onehot.loc[venues_onehot['Italian Restaurant'] == 0,["Neighborhood","Neighborhood Latitude","Neighborhood Longitude"]])



In [213]:
#Get a list of all the neighborhoods who have 0 italian restaurnts 
neighborhood_zero_italian_restaurant=get_neighborhoods_zero_italian_restaurants(All_venues).reset_index(drop = True)
neighborhood_zero_italian_restaurant.head(3)

Neighborhood  Neighborhood Latitude  Neighborhood Longitude
0     Allerton              40.865788              -73.859319
1    Arlington              40.635325              -74.165104
2     Arrochar              40.596313              -74.067124

## Visualisation:
-So now I have a list of all the Neighborhoods with zero italian restaurants, i will create a folium map for better demonstration.<br/>

In [210]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location= geolocator.geocode(address)
latitude_NewYork = location.latitude
longitude_NewYork = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude_NewYork, longitude_NewYork))

The geograpical coordinate of New York City are 40.7308619, -73.9871558.


In [211]:
# create map of New York using latitude and longitude of all the neighborhoods who have Zero Italian Restaurant
map_newyork = folium.Map(location=[latitude_NewYork, longitude_NewYork], zoom_start=10)

# add markers to map
for lat, lng, neighborhood in zip(neighborhood_zero_italian_restaurant['Neighborhood Latitude'], neighborhood_zero_italian_restaurant['Neighborhood Longitude'], neighborhood_zero_italian_restaurant['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [215]:
#Number of neighborhoods with zero italian restaurants in New York
neighborhood_zero_italian_restaurant.shape

(239, 3)

## Conclusion:
-There are 239 neighborhoods in New York with Zero italian restaurants .<br/>
-First I will use this data with the FourSquare API to try to find the neighborhoods with Zero Italian Restaurants<br/>
-Then I will create a folium map of it and show it to the investor.<br/>